gdc_ng_models/models/cohort.py

```diff
@@ -0,0 +1,222 @@
+        updated_datetime: The date and time when the record is updated.
+    """
+
+    __tablename__ = "cohort_filter"
+    id = Column(BigInteger, primary_key=True)
+    parent_id = Column(BigInteger, ForeignKey("cohort_filter.id"))
+    cohort_id = Column(UUID(as_uuid=True), ForeignKey("cohort.id"), nullable=False)
+    filters = Column(JSONB, nullable=False)
+    static = Column(Boolean, nullable=False, default=False)
+
+    # establishes a one-to-many relationship with CohortSnapshot
+    snapshots = orm.relationship("CohortSnapshot")
+
+    # establishes an adjacency relationship (i.e. self-referential key)
+    parent = orm.relationship("CohortFilter")
+
+    def __repr__(self):
+        return (
+            "<CohortFilter("
+            "id={id}, "
+            "parent_id={parent_id}, "
+            "cohort_id={cohort_id}, "
+            "filters={filters}, "
+            "static={static}, "
+            "created_datetime={created_datetime}, "
+            "updated_datetime={updated_datetime})>".format(
+                id=self.id,
+                parent_id=self.parent_id,
+                cohort_id=self.cohort_id,
+                filters=self.filters,
+                static=self.static,
+                created_datetime=self.created_datetime.isoformat() if self.created_datetime else None,
+                updated_datetime=self.updated_datetime.isoformat() if self.updated_datetime else None,
+            )
+        )
+
+    def to_json(self):
+        return {
+            "id": self.id,
+            "parent_id": self.parent_id,
+            "cohort_id": str(self.cohort_id),
+            "filters": self.filters,
+            "static": self.static,
+            "created_datetime": self.created_datetime.isoformat() if self.created_datetime else None,
+            "updated_datetime": self.updated_datetime.isoformat() if self.updated_datetime else None,
+        }
+
+
+class CohortSnapshot(Base, audit.AuditColumnsMixin):
+    """A static snapshot of cases associated with a cohort filter.
+
+    In the event a cohort is static, this provides a list of cases associated
+    with the cohort and the data release in which they were generated.
+
+    Attributes:
+        id: A bigint identifier for the snapshot.
+        filter_id: The ID of the filter associated with the snapshot.
+        data_release: The UUID of the data release when the snapshot was created.
+        case_ids: A UUID array containing the set of case IDs.
+        created_datetime: The date and time when the record is created.
+        updated_datetime: The date and time when the record is updated.
+    """
+
+    __tablename__ = "cohort_snapshot"
+    id = Column(BigInteger, primary_key=True)
+    filter_id = Column(BigInteger, ForeignKey("cohort_filter.id"), nullable=False)
+    data_release = Column(UUID(as_uuid=True), nullable=False)
+    case_ids = Column(ARRAY(UUID(as_uuid=True)), nullable=False)
+
```

Comment:
do we need to define the orm.relationship between CohortSnapshot and CohortFilter?

Reply:
I'm curious to know that too. It is my understanding that it is an 'either or' strategy. Either you have a static set or you use a filter, but may preserve the filter that built the static set?

Reply:
It wasn't necessary for how it's currently setup but when we add in the bi-directional relationships using back_populate then we'll have to specify the relationship on both sides. Those will be in the next set of commits.

Reply:
The updates to include bidirectional references are complete.